File: androidp1/app/src/main/java/com/preappointment1/app/ui/screens/DashboardScreen.kt
package com.preappointment1.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.preappointment1.app.R
import com.preappointment1.app.data.model.AgentResponse
import com.preappointment1.app.data.model.SubscriptionResponse
import com.preappointment1.app.data.model.TimelineEventResponse
import com.preappointment1.app.ui.components.*
import com.preappointment1.app.ui.support.FileReadiness
import com.preappointment1.app.ui.support.FileStats
import com.preappointment1.app.ui.theme.*
import java.time.Instant
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale

data class FollowUpUi(
    val id: String,
    val title: String,
    val daysRemaining: Int,
    val totalDays: Int,
    val progress: Float,
    val isActive: Boolean,
    val rules: com.preappointment1.app.data.model.FollowUpRules?,
    val schedule: Map<String, List<String>>?,
    val startsAt: String = "",
    val expiresAt: String = ""
)

@Composable
fun DashboardScreen(
    followUps: List<FollowUpUi>,
    timelineByFollowUpId: Map<String, List<TimelineEventResponse>>,
    patientName: String?,
    isLoading: Boolean,
    onNewFollowUp: () -> Unit,
    onOpenJourney: (FollowUpUi) -> Unit,
    onOpenNotifications: () -> Unit,
    modifier: Modifier = Modifier
) {
    var showComingSoon by remember { mutableStateOf(false) }

    if (showComingSoon) {
        ComingSoonDialog(onDismiss = { showComingSoon = false })
    }

    val activeFollowUps = followUps.filter { it.isActive }
    val greeting = FileStats.greetingPrefix()
    val firstName = patientName?.trim()?.split(" ")?.firstOrNull()?.takeIf { it.isNotBlank() }

    Scaffold(
        containerColor = White,
        modifier = modifier
    ) { padding ->
        if (isLoading) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding),
                contentAlignment = Alignment.Center
            ) {
                CircularProgressIndicator(color = Black)
            }
        } else if (activeFollowUps.isEmpty()) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding),
                contentAlignment = Alignment.Center
            ) {
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(32.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Text(
                        text = stringResource(R.string.dashboard_empty_title),
                        style = MaterialTheme.typography.titleLarge,
                        fontWeight = FontWeight.Black,
                        color = Black,
                        textAlign = TextAlign.Center
                    )
                    Spacer(modifier = Modifier.height(16.dp))
                    Text(
                        text = stringResource(R.string.dashboard_empty_body),
                        style = MaterialTheme.typography.bodyMedium,
                        color = Gray600,
                        textAlign = TextAlign.Center,
                        lineHeight = 24.sp
                    )
                    Spacer(modifier = Modifier.height(32.dp))
                    LpmPrimaryButton(
                        text = stringResource(R.string.dashboard_empty_action),
                        onClick = onNewFollowUp
                    )
                }
            }
        } else {
            LazyColumn(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding),
                contentPadding = PaddingValues(top = 8.dp, bottom = 32.dp)
            ) {
                item {
                    Column(modifier = Modifier.padding(horizontal = 24.dp, vertical = 16.dp)) {
                        Text(
                            text = if (firstName != null) {
                                stringResource(R.string.dashboard_greeting_named, greeting, firstName)
                            } else {
                                stringResource(R.string.dashboard_greeting, greeting)
                            },
                            style = MaterialTheme.typography.headlineSmall,
                            fontWeight = FontWeight.Black,
                            color = Black
                        )
                        Spacer(modifier = Modifier.height(4.dp))
                        Text(
                            text = stringResource(R.string.dashboard_subtitle),
                            style = MaterialTheme.typography.bodyMedium,
                            color = Gray600
                        )
                    }
                }

                items(activeFollowUps) { followUp ->
                    val events = timelineByFollowUpId[followUp.id] ?: emptyList()
                    val readiness = FileStats.compute(followUp, events)
                    FollowUpFileCard(
                        followUp = followUp,
                        readiness = readiness,
                        onClick = { onOpenJourney(followUp) }
                    )
                }
            }
        }
    }
}

@Composable
private fun FollowUpFileCard(
    followUp: FollowUpUi,
    readiness: FileReadiness,
    onClick: () -> Unit
) {
    LpmCard(
        modifier = Modifier.padding(horizontal = 24.dp, vertical = 8.dp),
        onClick = onClick
    ) {
        Column(modifier = Modifier.padding(20.dp)) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.Top
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = followUp.title,
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold,
                        color = Black
                    )
                    Spacer(modifier = Modifier.height(4.dp))
                    Text(
                        text = if (followUp.daysRemaining == 0) {
                            stringResource(R.string.file_appt_today)
                        } else {
                            stringResource(R.string.file_appt_countdown, followUp.daysRemaining)
                        },
                        style = MaterialTheme.typography.bodySmall,
                        color = Gray600
                    )
                }
                Text(
                    text = stringResource(R.string.file_ready_percent, readiness.readinessPercent),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Black,
                    color = Black
                )
            }

            Spacer(modifier = Modifier.height(12.dp))
            LpmProgressBar(progress = readiness.readinessPercent / 100f)

            Spacer(modifier = Modifier.height(12.dp))
            Text(
                text = stringResource(
                    R.string.file_contents,
                    readiness.measurementCount,
                    readiness.photoCount,
                    readiness.noteCount
                ),
                style = MaterialTheme.typography.bodySmall,
                color = Gray600
            )

            if (readiness.dayStreak >= 2) {
                Spacer(modifier = Modifier.height(8.dp))
                Text(
                    text = stringResource(R.string.file_streak, readiness.dayStreak),
                    style = MaterialTheme.typography.labelMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = Black
                )
            }

            readiness.missingHint?.let { hint ->
                Spacer(modifier = Modifier.height(8.dp))
                Text(
                    text = hint,
                    style = MaterialTheme.typography.bodySmall,
                    color = Gray600,
                    lineHeight = 20.sp
                )
            }

            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = readiness.tagline,
                style = MaterialTheme.typography.bodySmall,
                color = Black,
                fontWeight = FontWeight.Medium,
                lineHeight = 20.sp
            )

            Spacer(modifier = Modifier.height(16.dp))
            Text(
                text = stringResource(R.string.file_continue),
                style = MaterialTheme.typography.labelLarge,
                fontWeight = FontWeight.Bold,
                color = Black
            )
        }
    }
}

@Composable
fun ComingSoonDialog(onDismiss: () -> Unit) {
    AlertDialog(
        onDismissRequest = onDismiss,
        containerColor = White,
        shape = RoundedCornerShape(4.dp),
        title = {
            Text(
                "Coming soon",
                fontWeight = FontWeight.Bold,
                color = Black
            )
        },
        text = {
            Text(
                "Scanning prescriptions and doctor protocols will be available in a future update.",
                color = Gray600
            )
        },
        confirmButton = {
            TextButton(onClick = onDismiss) {
                Text("OK", color = Black, fontWeight = FontWeight.SemiBold)
            }
        }
    )
}

fun SubscriptionResponse.toFollowUpUi(agents: Map<String, AgentResponse>): FollowUpUi {
    val start = runCatching { Instant.parse(starts_at) }.getOrNull() ?: Instant.now()
    val end = runCatching { Instant.parse(expires_at) }.getOrNull() ?: start.plus(14, ChronoUnit.DAYS)
    val now = Instant.now()
    val totalDays = ChronoUnit.DAYS.between(start, end).coerceAtLeast(1).toInt()
    val elapsedDays = ChronoUnit.DAYS.between(start, now).coerceAtLeast(0)
    val daysRemaining = ChronoUnit.DAYS.between(now, end).coerceAtLeast(0).toInt()
    val progress = (elapsedDays.toFloat() / totalDays.toFloat()).coerceIn(0f, 1f)
    val startDateStr = start.toString().take(10)
    val agent = agents[agent_id]
    val title = parameters?.get("title")?.toString() ?: agent?.name ?: "Tracking from $startDateStr"

    var parsedRules: com.preappointment1.app.data.model.FollowUpRules? = null
    val rulesMap = parameters?.get("rules") as? Map<*, *>
    if (rulesMap != null) {
        parsedRules = com.preappointment1.app.data.model.FollowUpRules(
            temperature = rulesMap["temperature"] as? Boolean ?: false,
            pain = rulesMap["pain"] as? Boolean ?: false,
            photos = rulesMap["photos"] as? Boolean ?: false,
            smartwatch = rulesMap["smartwatch"] as? Boolean ?: false,
            bloodPressure = rulesMap["blood_pressure"] as? Boolean ?: false
        )
    }

    var parsedSchedule: Map<String, List<String>>? = null
    val scheduleMap = parameters?.get("schedule") as? Map<*, *>
    if (scheduleMap != null) {
        parsedSchedule = scheduleMap.mapNotNull { (k, v) ->
            val key = k as? String
            val valueList = (v as? List<*>)?.mapNotNull { it as? String }
            if (key != null && valueList != null) key to valueList else null
        }.toMap()
    }

    return FollowUpUi(
        id = id,
        title = title,
        daysRemaining = daysRemaining,
        totalDays = totalDays,
        progress = progress,
        isActive = now.isBefore(end),
        rules = parsedRules,
        schedule = parsedSchedule,
        startsAt = starts_at,
        expiresAt = expires_at
    )
}

File: iosp1/P1/P1/UI/Screens/DashboardScreen.swift
import SwiftUI

struct FollowUpUi: Identifiable, Equatable {
    let id: String
    let title: String
    let daysRemaining: Int
    let totalDays: Int
    let progress: Float
    let isActive: Bool
    let startsAt: String
    let expiresAt: String
    let rules: FollowUpRules?
    let schedule: [String: [String]]?
}

struct DashboardScreen: View {
    let onNewFollowUp: () -> Void
    let onOpenJourney: (FollowUpUi) -> Void
    let onOpenNotifications: () -> Void
    let onOpenDrawer: () -> Void

    @State private var followUps: [FollowUpUi] = []
    @State private var timelineByFollowUpId: [String: [TimelineEventResponse]] = [:]
    @State private var isLoading = true
    @State private var errorMessage: String?
    @State private var isOfflineMode = false
    @State private var pendingSyncCount = 0

    private var greeting: String { FileStats.greetingPrefix() }

    private var firstName: String? {
        SessionManager.shared.getUserName()?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .split(separator: " ")
            .first
            .map(String.init)
    }

    var body: some View {
        ZStack {
            Color.white.ignoresSafeArea()

            VStack(spacing: 0) {
                HStack {
                    Button(action: onOpenDrawer) {
                        Image(systemName: "line.3.horizontal")
                            .font(.title2)
                            .foregroundColor(.black)
                    }

                    Text("P1")
                        .font(.title2)
                        .fontWeight(.bold)
                        .padding(.leading, 8)
                    
                    if isOfflineMode || pendingSyncCount > 0 {
                        Text(isOfflineMode ? "Offline" : "\(pendingSyncCount) pending")
                            .font(.caption2)
                            .foregroundColor(Color(UIColor.systemGray))
                    }
                    
                    Spacer()

                    Button(action: onOpenNotifications) {
                        Image(systemName: "bell")
                            .font(.title3)
                            .foregroundColor(.black)
                    }
                }
                .padding()

                content
            }
        }
        .task { await loadData() }
    }

    @ViewBuilder
    private var content: some View {
        if isLoading {
            Spacer()
            ProgressView()
                .progressViewStyle(CircularProgressViewStyle(tint: .black))
                .scaleEffect(1.5)
            Spacer()
        } else if let error = errorMessage {
            Spacer()
            VStack(spacing: 16) {
                Text(error)
                    .foregroundColor(.black)
                    .multilineTextAlignment(.center)
                LpmPrimaryButton(text: "Retry") { Task { await loadData() } }
            }
            .padding()
            Spacer()
        } else if followUps.filter({ $0.isActive }).isEmpty {
            Spacer()
            VStack(spacing: 16) {
                Text(String(localized: "welcome_title"))
                    .font(.title2)
                    .fontWeight(.black)
                    .multilineTextAlignment(.center)

                Text(String(localized: "welcome_desc"))
                    .font(.body)
                    .foregroundColor(Color(UIColor.systemGray))
                    .multilineTextAlignment(.center)
                    .padding(.horizontal, 32)

                LpmPrimaryButton(text: String(localized: "welcome_start"), action: onNewFollowUp)
            }
            .padding()
            Spacer()
        } else {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    VStack(alignment: .leading, spacing: 4) {
                        if let name = firstName {
                            Text("\(greeting), \(name)")
                                .font(.title2)
                                .fontWeight(.black)
                        } else {
                            Text(greeting)
                                .font(.title2)
                                .fontWeight(.black)
                        }
                        Text(String(localized: "dashboard_subtitle"))
                            .font(.body)
                            .foregroundColor(Color(UIColor.systemGray))
                    }
                    .padding(.horizontal, 24)
                    .padding(.vertical, 16)

                    ForEach(followUps.filter { $0.isActive }) { followUp in
                        let events = timelineByFollowUpId[followUp.id] ?? []
                        FollowUpFileCard(
                            followUp: followUp,
                            readiness: FileStats.compute(followUp: followUp, events: events),
                            onTap: { onOpenJourney(followUp) }
                        )
                        .padding(.horizontal, 24)
                        .padding(.vertical, 8)
                    }
                }
                .padding(.bottom, 32)
            }
        }
    }

    private func loadData() async {
        isLoading = true
        errorMessage = nil
        
        guard SessionManager.shared.getToken() != nil else {
            errorMessage = "Unable to connect. Check your network."
            followUps = []
            isLoading = false
            return
        }
        
        let result = await FollowUpRepository.loadFollowUpsWithSync()
        followUps = result.followUps
        isOfflineMode = !result.synced
        
        var timelines: [String: [TimelineEventResponse]] = [:]
        for followUp in followUps.filter({ $0.isActive }) {
            timelines[followUp.id] = TimelineRepository.getEvents(subscriptionId: followUp.id)
        }
        timelineByFollowUpId = timelines
        pendingSyncCount = LocalStore.shared.pendingCount()
        
        isLoading = false
    }
}

private struct FollowUpFileCard: View {
    let followUp: FollowUpUi
    let readiness: FileReadiness
    let onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            VStack(alignment: .leading, spacing: 12) {
                HStack(alignment: .top) {
                    VStack(alignment: .leading, spacing: 4) {
                        Text(followUp.title)
                            .font(.headline)
                            .fontWeight(.bold)
                            .foregroundColor(.black)
                        Text(
                            followUp.daysRemaining == 0
                                ? String(localized: "file_appt_today")
                                : String(format: String(localized: "file_appt_countdown"), followUp.daysRemaining)
                        )
                        .font(.caption)
                        .foregroundColor(Color(UIColor.systemGray))
                    }
                    Spacer()
                    Text(String(format: String(localized: "file_ready_percent"), readiness.readinessPercent))
                        .font(.headline)
                        .fontWeight(.black)
                        .foregroundColor(.black)
                }

                ProgressView(value: Double(readiness.readinessPercent), total: 100)
                    .tint(.black)

                Text(String(
                    format: String(localized: "summary_file_contents"),
                    readiness.measurementCount,
                    readiness.photoCount,
                    readiness.noteCount
                ))
                .font(.caption)
                .foregroundColor(Color(UIColor.systemGray))

                if readiness.dayStreak >= 2 {
                    Text(String(format: String(localized: "file_streak"), readiness.dayStreak))
                        .font(.caption)
                        .fontWeight(.semibold)
                        .foregroundColor(.black)
                }

                if let hint = readiness.missingHint {
                    Text(hint)
                        .font(.caption)
                        .foregroundColor(Color(UIColor.systemGray))
                }

                Text(readiness.tagline)
                    .font(.caption)
                    .foregroundColor(.black)
                    .fontWeight(.medium)

                Text(String(localized: "file_continue"))
                    .font(.subheadline)
                    .fontWeight(.bold)
                    .foregroundColor(.black)
                    .padding(.top, 4)
            }
            .padding(20)
            .background(Color.white)
            .overlay(
                RoundedRectangle(cornerRadius: 4)
                    .stroke(Color(UIColor.systemGray5), lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
    }
}
